// fuse-box/dist/commonjs/Log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ansi = require("ansi");
const cursor = ansi(process.stdout);
const prettysize = require("prettysize");
const prettyTime = require("pretty-time");
class Log {
    constructor(context) {
        this.context = context;
        this.timeStart = process.hrtime();
        this.printLog = true;
        this.totalSize = 0;
        this.printLog = context.doLog;
    }
    reset() {
        this.timeStart = process.hrtime();
        this.totalSize = 0;
    }
    echoWith(str, opt) {
        if (this.printLog) {
            cursor.write(` `)[opt]().write(str);
            cursor.write("\n");
            cursor.reset();
        }
    }
    bundleStart(name) {
        if (this.printLog) {
            cursor.reset()
                .cyan().bold().write("\n" + name + " -> \n\n");
            cursor.reset();
        }
    }
    subBundleStart(name, parent) {
        if (this.printLog) {
            cursor.reset()
                .cyan().write("\n" + name + ` (child of ${parent}) -> \n\n`);
            cursor.reset();
        }
    }
    bundleEnd(name, collection) {
        if (this.printLog) {
            let took = process.hrtime(this.timeStart);
            cursor.reset().write(`-> Finished `)
                .green().write(name)
                .green().write(` ${collection.cachedName || collection.name}`)
                .yellow().write(`took: ${prettyTime(took, "ms")}`);
        }
    }
    echoHeader(str) {
        if (this.printLog) {
            cursor.write(` `).yellow().write(str);
            cursor.write("\n");
            cursor.reset();
        }
    }
    echo(str) {
        if (this.printLog) {
            let data = new Date();
            let hour = data.getHours();
            let min = data.getMinutes();
            let sec = data.getSeconds();
            hour = hour < 10 ? `0${hour}` : hour;
            min = min < 10 ? `0${min}` : min;
            sec = sec < 10 ? `0${sec}` : sec;
            cursor.yellow().write(`${hour}:${min}:${sec} : `)
                .green().write(str);
            cursor.write("\n");
            cursor.reset();
        }
    }
    echoStatus(str) {
        if (this.printLog) {
            cursor.write(`  → `)
                .cyan().write(str);
            cursor.write("\n");
            cursor.reset();
        }
    }
    echoWarning(str) {
        cursor.red().write(`  → WARNING `)
            .write(str);
        cursor.write("\n");
        cursor.reset();
    }
    echoDefaultCollection(collection, contents) {
        if (!this.printLog) {
            return;
        }
        let bytes = Buffer.byteLength(contents, "utf8");
        let size = prettysize(bytes);
        this.totalSize += bytes;
        cursor.reset().write(`└──`)
            .green().write(` ${collection.cachedName || collection.name}`)
            .yellow().write(` (${collection.dependencies.size} files,  ${size})`);
        cursor.write("\n");
        collection.dependencies.forEach(file => {
            if (!file.info.isRemoteFile) {
                cursor.reset().write(`      ${file.info.fuseBoxPath}`).write("\n");
            }
        });
        cursor.reset();
    }
    echoCollection(collection, contents) {
        if (!this.printLog) {
            return;
        }
        let bytes = Buffer.byteLength(contents, "utf8");
        let size = prettysize(bytes);
        this.totalSize += bytes;
        cursor.reset().write(`└──`)
            .green().write(` ${collection.cachedName || collection.name}`)
            .reset().write(` (${collection.dependencies.size} files)`)
            .yellow().write(` ${size}`)
            .write("\n").reset();
    }
    end(header) {
        let took = process.hrtime(this.timeStart);
        this.echoBundleStats(header || "Bundle", this.totalSize, took);
    }
    echoBundleStats(header, size, took) {
        if (!this.printLog) {
            return;
        }
        cursor.write("\n")
            .yellow().write(`    Size: ${prettysize(size)}`)
            .write(` in ${prettyTime(took, "ms")}`)
            .write("\n").reset();
    }
}
exports.Log = Log;
